### PR TITLE
Added UpdatePaymentInstrument mutation

### PR DESCRIPTION
**Pull request by nikithaguduru**
Added UpdatePaymentInstrument mutation

## Summary by Sourcery

Add a new GraphQL mutation for updating payment instrument details, including necessary input types and API enhancements.

New Features:
- Introduce UpdatePaymentInstrument mutation to allow updating payment instrument details for a member.

Enhancements:
- Add UpdatePaymentInstrumentInput type to define the input structure for the new mutation.
- Extend the PaymentCybersourceApi interface and its implementation to support updating payment instruments.

### CODE REVIEW

File: data-access/src/routes/http/graphql/schema/builder/graphql.schema.json [hunk 1]
```diff
@@ -14655,6 +14688,269 @@
         "enumValues": null,
         "possibleTypes": null
       },
+      {
+        "kind": "INPUT_OBJECT",
+        "name": "UpdatePaymentInstrumentInput",
```
[the remark this answers is not in the document]
Comment (heruwala) [reply data-access/src/routes/http/graphql/schema/builder/graphql.schema.json:14693]: Not relevant as the expected values are in string

File: data-access/src/routes/http/graphql/schema/builder/graphql.schema.json [hunk 2]
```diff
@@ -14655,6 +14688,269 @@
...
+                "name": "String",
+                "ofType": null
+              }
+            },
+            "defaultValue": null,
+            "isDeprecated": false,
+            "deprecationReason": null
+          },
+          {
+            "name": "cardType",
```
[the remark this answers is not in the document]
Comment (heruwala) [reply data-access/src/routes/http/graphql/schema/builder/graphql.schema.json:14854]: could you implement an enum to restrict the card types (e.g. 001, 002, 003 and 004)

File: data-access/src/routes/http/graphql/schema/types/member.resolvers.ts [hunk 3]
```diff
@@ -208,7 +208,33 @@ const member: Resolvers = {
         return { status: { success: true }, member };
       }
     },
-
+    memberUpdatePaymentInstrument: async (_, { input }, context) => {
+      const member = await getMemberForCurrentUser(context);
+      try {
+        let cybersourceCustomerId = member?.cybersourceCustomerId;
+        if (cybersourceCustomerId) {
+          const paymentInstrumentPayload: CustomerProfile = {
```
[the remark this answers is not in the document]
Comment (heruwala) [reply data-access/src/routes/http/graphql/schema/types/member.resolvers.ts:216]: let's implement this feedback.

File: data-access/src/routes/http/graphql/schema/types/member.resolvers.ts [hunk 4]
```diff
@@ -208,7 +208,33 @@ const member: Resolvers = {
...
+          };
+          const paymentInfo: PaymentInstrumentInfo = {
+            id: input.id,
+            paymentInstrumentId: input.paymentInstrumentId,
+            cardExpirationMonth: input.expirationMonth,
+            cardExpirationYear: input.expirationYear,
+            cardType: input.cardType            
+          }
+          const response = await context.applicationServices.member.cybersourceApi.updatePaymentInstrument(paymentInstrumentPayload, paymentInfo);
+          if (response && input.isDefault) {
```
Comment (heruwala) [inline data-access/src/routes/http/graphql/schema/types/member.resolvers.ts:228]: Add an extra check here to the condtion response.Active

File: data-access/src/routes/http/graphql/schema/types/member.resolvers.ts [hunk 5]
```diff
@@ -208,7 +208,33 @@ const member: Resolvers = {
...
+            await context.applicationServices.member.cybersourceApi.setDefaultPaymentInstrument(cybersourceCustomerId, paymentInfo.paymentInstrumentId);
+            return { status: {success: true}, member };   
+          }
+          return { status: {success: false}, member };       
+      }
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (heruwala) [reply data-access/src/routes/http/graphql/schema/types/member.resolvers.ts:234]: we will revisit all error handling as a whole for violation tickets, later.

File: data-access/src/app/application-services/member/member.payment.ts [hunk 6]
```diff
@@ -110,6 +103,16 @@ export class PaymentCybersourceApiImpl extends PaymentDataSource<AppContext> imp
     return response;
   }
 
+  public async updatePaymentInstrument(customerProfile: CustomerProfile, paymentInstrumentInfo: PaymentInstrumentInfo): Promise<boolean> {
+    let cyberSourcePaymentInstrumentsResponse: any;
```
Comment (heruwala) [inline data-access/src/app/application-services/member/member.payment.ts:107]: use proper types instead of any.

**Review by heruwala — CHANGES_REQUESTED** [review]
few minor issues, please fix and test them out

File: data-access/src/routes/http/graphql/schema/types/member.resolvers.ts [hunk 7]
```diff
@@ -208,7 +208,42 @@ const member: Resolvers = {
...
+            cardExpirationMonth: input.expirationMonth,
+            cardExpirationYear: input.expirationYear,
+            cardType: input.cardType            
+          }
+          const response = await context.applicationServices.member.cybersourceApi.updatePaymentInstrument(paymentInstrumentPayload, paymentInfo);
+          if (response && input.isDefault) {
+            await context.applicationServices.member.cybersourceApi.setDefaultPaymentInstrument(cybersourceCustomerId, paymentInfo.paymentInstrumentId);
+            return { status: {success: true}, member };   
+          }
+          return { status: {success: false}, member };
```
Comment (heruwala) [inline data-access/src/routes/http/graphql/schema/types/member.resolvers.ts:241]: Why this line is returning success: false here? What if someone updated payment instrument and did not ask for default change or vice versa?

File: data-access/src/routes/http/graphql/schema/types/member.graphql [hunk 8]
```diff
@@ -199,3 +200,30 @@ input AddPaymentInstrumentInput {
...
+  billingPostalCode: String!
+  billingCountry: String!
+}
+
+enum CardType {
+  Visa
+  Master
+  American_Express
+  Discover
+  JCB
```
Comment (heruwala) [inline data-access/src/routes/http/graphql/schema/types/member.graphql:228]: we do not accept JCB, keep the enum limited to Visa, Master Card, Amex and Discover. Also, please verify the accepted values for cybersource.

**Comment by heruwala** [comment]
@sourcery-ai review

File: data-access/src/app/application-services/member/member.payment.ts [hunk 9]
```diff
@@ -110,6 +103,16 @@ export class PaymentCybersourceApiImpl extends PaymentDataSource<AppContext> imp
     return response;
   }
 
+  public async updatePaymentInstrument(customerProfile: CustomerProfile, paymentInstrumentInfo: PaymentInstrumentInfo): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (heruwala) [reply data-access/src/app/application-services/member/member.payment.ts:106]: not necessary per current implementation

**Comment by heruwala** [comment]
Please review the sourcery feedback as well as few changes requested by me.

**Comment by heruwala** [comment]
@sourcery-ai review

File: data-access/src/app/application-services/member/member.payment.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (nikithaguduru) [reply data-access/src/app/application-services/member/member.payment.ts:106]: ignoring this due to complex nature of the process

File: data-access/src/routes/http/graphql/schema/builder/graphql.schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikithaguduru) [reply data-access/src/routes/http/graphql/schema/builder/graphql.schema.json:14878]: enums are not possible for this integration

File: data-access/src/routes/http/graphql/schema/types/member.graphql [hunk 10]
```diff
@@ -199,3 +200,30 @@ input AddPaymentInstrumentInput {
   paymentToken: String!
   isDefault: Boolean!
 }
+
+input UpdatePaymentInstrumentInput {
```
[the remark this answers is not in the document]
Comment (nikithaguduru) [reply data-access/src/routes/http/graphql/schema/types/member.graphql:204]: not valid as enum is not being used

**Review by heruwala — APPROVED** [review]

**Pull request merged**